src/group_theory/order_of_element.lean

```diff
@@ -8,388 +8,364 @@ import group_theory.coset
-  `a` has infinite order.
+* `is_of_fin_order` is a predicate on an element `x` of a monoid `G` saying that `x` is of finite
+  order.
+* `is_of_fin_add_order` is the additive analogue of `is_of_find_order`.
+* `order_of x` defines the order of an element `x` of a monoid `G`, by convention its value is `0`
+  if `x` has infinite order.
+* `add_order_of` is the additive analogue of `order_of`.
 
 ## Tags
-
 order of an element
+-/
 
-## TODO
+open function nat
 
-* Move the first declarations until the definition of order to other files.
-* Yury's suggestion: Redefine `order_of x := minimal_period (* x) 1`, this should make `to_additive`
-  easier.
--/
+universes u v
 
-open function
-open_locale big_operators
+variables {G : Type u} {A : Type v}
+variables {x : G} {a : A} {n m : ℕ}
 
-universe u
+section monoid_add_monoid
 
-variables {α : Type u} {s : set α} {a a₁ a₂ b c: α}
+variables [monoid G] [add_monoid A]
 
-section order_of
+section is_of_fin_order
 
-section monoid
-variables {α} [monoid α]
-variables {H : Type u} [add_monoid H] {x : H}
+lemma is_periodic_pt_add_iff_nsmul_eq_zero (a : A) :
+  is_periodic_pt ((+) a) n 0 ↔ n •ℕ a = 0 :=
+by rw [is_periodic_pt, is_fixed_pt, add_left_iterate, add_zero]
 
-open_locale classical
+@[to_additive is_periodic_pt_add_iff_nsmul_eq_zero]
+lemma is_periodic_pt_mul_iff_pow_eq_one (x : G) : is_periodic_pt ((*) x) n 1 ↔ x ^ n = 1 :=
+by rw [is_periodic_pt, is_fixed_pt, mul_left_iterate, mul_one]
 
-/-- `add_order_of h` is the additive order of the element `x`, i.e. the `n ≥ 1`, s. t. `n •ℕ x = 0`
-if it exists. Otherwise, i.e. if `x` is of infinite order, then `add_order_of x` is `0` by
-convention.-/
-noncomputable def add_order_of (x : H) : ℕ :=
-if h : ∃ n, 0 < n ∧ n •ℕ x = 0 then nat.find h else 0
+/-- `is_of_fin_add_order` is a predicate on an element `a` of an additive monoid to be of finite
+order, i.e. there exists `n ≥ 1` such that `n •ℕ a = 0`.-/
+def is_of_fin_add_order (a : A) : Prop :=
+(0 : A) ∈ periodic_pts ((+) a)
 
-/-- `order_of a` is the order of the element `a`, i.e. the `n ≥ 1`, s.t. `a ^ n = 1` if it exists.
-Otherwise, i.e. if `a` is of infinite order, then `order_of a` is `0` by convention.-/
-noncomputable def order_of (a : α) : ℕ :=
-if h : ∃ n, 0 < n ∧ a ^ n = 1 then nat.find h else 0
+/-- `is_of_fin_order` is a predicate on an element `x` of a monoid to be of finite order, i.e. there
+exists `n ≥ 1` such that `x ^ n = 1`.-/
+@[to_additive is_of_fin_add_order]
+def is_of_fin_order (x : G) : Prop :=
+(1 : G) ∈ periodic_pts ((*) x)
 
-attribute [to_additive add_order_of] order_of
+lemma is_of_fin_add_order_of_mul_iff :
+is_of_fin_add_order (additive.of_mul x) ↔ is_of_fin_order x :=
+by refl
 
-@[simp] lemma add_order_of_of_mul_eq_order_of (a : α) :
-  add_order_of (additive.of_mul a) = order_of a :=
-by simp [add_order_of, order_of, ← of_mul_pow]
+lemma is_of_fin_order_of_add_iff :
+is_of_fin_order (multiplicative.of_add a) ↔ is_of_fin_add_order a :=
+by refl
```

Comment:
```suggestion
lemma is_of_fin_order_of_add_iff :
  is_of_fin_order (multiplicative.of_add a) ↔ is_of_fin_add_order a := iff.rfl
```

Reply:
Ah, `iff.rfl`, of course!  :upside_down_face: :left_right_arrow: